### ISSUE TITLE
"Failed to send message: Unable to parse entity" error on HASS

### ISSUE THREAD

**Issue by jprates** [bbernhard/signal-cli-rest-api#151]
### The problem

I've been using Signal Messenger 0.41 add-on for Home Assistant with "native" and "auto-receive" options active for several months with no errors whatsoever.

Suddenly since earlier today (2021-07-20 01:23 Lisbon time) some messages pass, others fail with this error:

`2021-07-20 01:22:37 ERROR (SyncWorker_10) [homeassistant.components.signal_messenger.notify] Failed to send message: Unable to parse entity`

Can't make out why some work and some don't, since they are all being sent to the same people.

Please let me know how can I assist in debugging this issue.

### Are you using the latest released version?

- [X] Yes

### Have you read the troubleshooting page?

- [X] Yes

### What type of installation are you running?

Home Assistant Signal Addon

### In which mode are you using the docker container?

Native Mode

### What's the architecture of your host system?

x86-64

### Additional information

**Core log:**

`2021-07-20 19:59:53 ERROR (SyncWorker_16) [homeassistant.components.signal_messenger.notify] Failed to send message: Unable to parse entity
2021-07-20 19:59:53 ERROR (MainThread) [homeassistant.core] Error executing service: <ServiceCall notify.signal_casa (c:2f49ced76d6935274232ad73c7d1ff25): message=Template("Ambiente: Está bom tempo, aproveitar para arejar a casa."), data=attachments=['signal_attach/arejar_casa.jpg']>
Traceback (most recent call last):
File "/usr/src/homeassistant/homeassistant/core.py", line 1507, in catch_exceptions
await coro_or_task
File "/usr/src/homeassistant/homeassistant/core.py", line 1526, in _execute_service
await handler.job.target(service_call)
File "/usr/src/homeassistant/homeassistant/components/notify/__init__.py", line 159, in _async_notify_message_service
await self.async_send_message(**kwargs)
File "/usr/src/homeassistant/homeassistant/components/notify/__init__.py", line 136, in async_send_message
await self.hass.async_add_executor_job(
File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 52, in run
result = self.fn(*self.args, **self.kwargs)
File "/usr/src/homeassistant/homeassistant/components/signal_messenger/notify.py", line 79, in send_message
raise ex
File "/usr/src/homeassistant/homeassistant/components/signal_messenger/notify.py", line 76, in send_message
self._signal_cli_rest_api.send_message(message, self._recp_nrs, filenames)
File "/usr/local/lib/python3.9/site-packages/pysignalclirestapi/api.py", line 83, in send_message
raise exc
File "/usr/local/lib/python3.9/site-packages/pysignalclirestapi/api.py", line 79, in send_message
raise SignalCliRestApiError(json_resp["error"])
pysignalclirestapi.api.SignalCliRestApiError: Failed to send message: Unable to parse entity`

**Signal add-on log:**

`Native mode
1
AUTO RECEIVE SCHEDULE
0 22 * * *
+ set -e
+ [ -z /data ]
+ chown 1000:1000 -R /data
+ cat
+ cap_prefix=-cap_
+ cat /proc/sys/kernel/cap_last_cap
+ seq -s ,-cap_ 0 40
+ caps=-cap_0,-cap_1,-cap_2,-cap_3,-cap_4,-cap_5,-cap_6,-cap_7,-cap_8,-cap_9,-cap_10,-cap_11,-cap_12,-cap_13,-cap_14,-cap_15,-cap_16,-cap_17,-cap_18,-cap_19,-cap_20,-cap_21,-cap_22,-cap_23,-cap_24,-cap_25,-cap_26,-cap_27,-cap_28,-cap_29,-cap_30,-cap_31,-cap_32,-cap_33,-cap_34,-cap_35,-cap_36,-cap_37,-cap_38,-cap_39,-cap_40
+ exec setpriv --reuid=1000 --regid=1000 --init-groups --inh-caps=-cap_0,-cap_1,-cap_2,-cap_3,-cap_4,-cap_5,-cap_6,-cap_7,-cap_8,-cap_9,-cap_10,-cap_11,-cap_12,-cap_13,-cap_14,-cap_15,-cap_16,-cap_17,-cap_18,-cap_19,-cap_20,-cap_21,-cap_22,-cap_23,-cap_24,-cap_25,-cap_26,-cap_27,-cap_28,-cap_29,-cap_30,-cap_31,-cap_32,-cap_33,-cap_34,-cap_35,-cap_36,-cap_37,-cap_38,-cap_39,-cap_40 signal-cli-rest-api -signal-cli-config=/data
time="2021-07-20T19:38:15+01:00" level=info msg="Started Signal Messenger REST API"
[GIN] 2021/07/20 - 19:59:21 | 200 |     222.066µs |     172.30.32.1 | GET      "/v1/about"
[GIN] 2021/07/20 - 19:59:53 | 400 | 32.142741242s |     172.30.32.1 | POST     "/v2/send"`

Please note the 32s to send a picture 300x300 pixels is not normal at all, it usually gets sent quite fast.

**Comment by bbernhard** [comment]
That's really strange. 

Could you please try this here: https://github.com/bbernhard/signal-cli-rest-api/blob/master/doc/DEBUG.md

This helps to determine whether the errors are signal-cli related or caused by the REST API wrapper.

**Comment by jprates** [comment]
Hi,

Sorry for the delay, but I am sending a picture in attach, and the debug showed a different tmp file each time I ran, and the command failed because the tmp file was not there.

I suspect this happened because the image tmp file is created on the fly, and in debug mode we can't do that.

I worked around this by copying the file from the host directly into the container tmp dir and executing the command using that copied file, and it worked fine:

`➜  ~ docker cp config/signal_attach/arejar_casa.jpg b39a497b24c2a7edfb6926822dd8a960cb4668a456c7060e74429dfea24f2a66:/tmp/arejar_casa.jpg

➜  ~ docker exec -it b39a497b24c2a7edfb6926822dd8a960cb4668a456c7060e74429dfea24f2a66 /bin/bash

WARNING: signal-cli-rest-api runs as signal-api (not as root!)

Run 'su signal-api' before using signal-cli!

If you want to use signal-cli directly, don't forget to specify the config directory. e.g: "signal-cli --config /data"

root@1315902c-signal-messenger:/data/data# su signal-api

signal-api@1315902c-signal-messenger:/data/data$ echo 'Ambiente: Está bom tempo, aproveitar para arejar a casa.' | signal-cli-native --config /data -u +3519xxxxxxxx send +3519yyyyyyyy -a /tmp/arejar_casa.jpg

1626815121807

signal-api@1315902c-signal-messenger:/data/data$
`

**Comment by jprates** [comment]
@bbernhard did my post above give you all the info you needed?

I guess it showed that the message gets sent correctly via the signal-cli, right?

**Comment by bbernhard** [comment]
You are right, the image gets deleted (even in debug mode). I'll probably change that with the next release...this should make debugging a bit easier.

Is your issue reproducible? I mean, does the same text/image always fails via REST but always works via signal-cli?

**Comment by jprates** [comment]
Yes, I believe it is, I've been without any msg for the last day.

I was just going to try sending a new message via cli when I decided to search for the current container ID, and I got this unexpected result:

`➜  ~ docker images | grep signal`
`ghcr.io/haberda/signal_messenger/amd64            0.41        198e74499197   3 weeks ago    503MB`
`1315902c/amd64-addon-signal_messenger             0.36.1      2f00eca7059a   4 months ago   433MB`
`bbernhard/signal-cli-rest-api                     0.36        d50fc38ac425   4 months ago   433MB`
`bbernhard/signal-cli-rest-api                     0.34        5b479ea089ed   5 months ago   295MB`
`bbernhard/signal-cli-rest-api                     0.33        2bfad023667a   5 months ago   295MB`
`bbernhard/signal-cli-rest-api                     32          7cf4d48d7f29   6 months ago   295MB`
`bbernhard/signal-cli-rest-api                     31          d6b1a00a1c55   6 months ago   468MB`

Why is all this trash of previous versions filling up my docker images list?
And why is the path so different in the latest versions? Is this expected? Do I have the correct version at all?

I'm going to delete the previous versions of course, but I expected the upgrade method to do this for me.

**Comment by bbernhard** [comment]
> And why is the path so different in the latest versions?

The Home Assistant addon is maintained by @haberda, but as far as I know this is expected. If I remember correctly the Home Assistant addon originally was using the `bbernhard/signal-cli-rest-api` image directly. But as there were more and more modifications needed in order to properly integrate it into Home Assistant, @haberda created its own image based on `bbernhard/signal-cli-rest-api`. @haberda Please correct me if I am wrong here

> Why is all this trash of previous versions filling up my docker images list?

I am not using any addons with Home Assistant, so not sure if that's something Home Assistant takes care of or if the cleanup needs to be done manually by the user.

> Do I have the correct version at all?

You are using version `0.41`, so that looks good.

Regarding the issue:

* Does the issue also occur if you send a text-only message?
* Does it work if you send the same message to a different number?
* Could you maybe try disabling the native mode. Does it work then?

The `Unable to parse entity` error actually comes from `signal-cli`. But the strange thing is, that it does work when you run it manually. So either it's an issue that happens sporadically or there's something wrong in the way the REST API wrapper invokes the `signal-cli` binary.

**Comment by jprates** [comment]
Hi @bbernhard ,

Based on your answers I got even more concerned about having many signal-cli images on my HASS, it's not normal at all.

I went and removed all images and all containers, which was a bit troublesome because some of them had dependencies on one-another, not something I particulary like to see to be honest.

Note that even trying to uninstall the add-on from HASS was failing before deleting the containers and images by hand.
With all clean, both on docker side and HASS side, I went to re-install the add-on, which was a pain because I had to go through the process of registering the phone number again, which with captchas is not an easy task.

Right now all is working well again.

So my conclusion and recommendation is that perhaps you or @haberda should give some more attention to the way the upgrade process runs, and make sure each time a new release is out, the old one gets effectively replaced/deleted from HASS at docker level.

Thanks for your assistance, I'll close this issue now.

Cheers,
-jprates

**Comment by haberda** [comment]
@bbernhard  You are correct in that the add-on used to just be your container without modification, all I did was change the path to the data storage location. Now I build an new image based on your image so that I can insert options that enable native mode and/or auto-receive, etc. 

Anyway, in terms of the old images, prior to `0.36.1`, because I was directly using your image, hassio didn't delete old images reliably. However, after `0.36.1` the images clean-up just fine because I can build the image with the correct tags. Those old images are not needed and can be safely deleted. At the time when I switched to building my own images (~5-6 months ago) I added to the change log that the image clean-up was fixed.
